### PR TITLE
MGMT-16242: Suggest OCP images based on availability for architecture

### PR DESCRIPTION
**Pull request by paul-maidment**
When creating an infraenv, a validation is performed to ensure that the release images contain a compatible image for a given Openshift version.

This based on the MajorMinor version of The desired Openshift version and and MajorMinor version of the relase image Openshift version.

This is performed within the context of a specific architecture.

The error message presented to the user when no suitable release image could be isolated is not helpful and needs to contain more clues on how the user might resolve the issue.

This PR will list available Openshift versions supported in the release images for the user's architecture when this error occurs.

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by paul-maidment** [comment]
On quay.io/edge-infrastructure/assisted-service:latest
Uploaded a single custom manifest, started installation, then aborted

{code}
[root@rdu-infra-edge-18 assisted-test-infra]# curl $SERVICE_URI/api/assisted-install/v2/clusters/${CLUSTER_ID}/manifests | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    60  100    60    0     0   4615      0 --:--:-- --:--:-- --:--:--  4615
[
  {
    "file_name": "dummy-namespace.yaml",
    "folder": "manifests"
  }
]

[root@rdu-infra-edge-18 assisted-test-infra]# curl $SERVICE_URI/api/assisted-install/v2/clusters/${CLUSTER_ID}/manifests?include_system_generated=true | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   208  100   208    0     0  13866      0 --:--:-- --:--:-- --:--:-- 13866
[
  {
    "file_name": "dummy-namespace.yaml",
    "folder": "manifests"
  },
  {
    "file_name": "50-masters-chrony-configuration.yaml",
    "folder": "openshift"
  },
  {
    "file_name": "50-workers-chrony-configuration.yaml",
    "folder": "openshift"
  }
]
{code}

Now will immediately change the service image to quay.io/pmaidmen/assisted-service:VERIFY-MGMT-18165
Installation will be started, reset and then we will add another file

{code}
[root@rdu-infra-edge-18 assisted-test-infra]# curl $SERVICE_URI/api/assisted-install/v2/clusters/${CLUSTER_ID}/manifests | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    85  100    85    0     0   4722      0 --:--:-- --:--:-- --:--:--  4722
[
  {
    "file_name": "dummy-namespace.yaml",
    "folder": "manifests",
    "manifest_source": "user"
  }
]

[root@rdu-infra-edge-18 assisted-test-infra]# curl $SERVICE_URI/api/assisted-install/v2/clusters/${CLUSTER_ID}/manifests?include_system_generated=true | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   287  100   287    0     0  17937      0 --:--:-- --:--:-- --:--:-- 17937
[
  {
    "file_name": "dummy-namespace.yaml",
    "folder": "manifests",
    "manifest_source": "user"
  },
  {
    "file_name": "50-masters-chrony-configuration.yaml",
    "folder": "openshift",
    "manifest_source": "system"
  },
  {
    "file_name": "50-workers-chrony-configuration.yaml",
    "folder": "openshift",
    "manifest_source": "system"
  }
]
{code}

Assisted service upgraded to the new version (the one we are testing)
Now we shall start installation again and  reset the cluster

{code}
[root@rdu-infra-edge-18 assisted-test-infra]# curl $SERVICE_URI/api/assisted-install/v2/clusters/${CLUSTER_ID}/manifests?include_system_generated=true | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    85  100    85    0     0   4473      0 --:--:-- --:--:-- --:--:--  4473
[
  {
    "file_name": "dummy-namespace.yaml",
    "folder": "manifests",
    "manifest_source": "user"
  }
]
{code}

We will now add another user generated manifest and restart the installation

{code}
[root@rdu-infra-edge-18 assisted-test-infra]# curl $SERVICE_URI/api/assisted-install/v2/clusters/${CLUSTER_ID}/manifests | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   170  100   170    0     0   9444      0 --:--:-- --:--:-- --:--:--  9444
[
  {
    "file_name": "dummy-namespace-2.yaml",
    "folder": "manifests",
    "manifest_source": "user"
  },
  {
    "file_name": "dummy-namespace.yaml",
    "folder": "manifests",
    "manifest_source": "user"
  }
]
[root@rdu-infra-edge-18 assisted-test-infra]# curl $SERVICE_URI/api/assisted-install/v2/clusters/${CLUSTER_ID}/manifests?include_system_generated=true | jq                                                                                                                                                                                                                                
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   372  100   372    0     0  17714      0 --:--:-- --:--:-- --:--:-- 18600
[
  {
    "file_name": "dummy-namespace-2.yaml",
    "folder": "manifests",
    "manifest_source": "user"
  },
  {
    "file_name": "dummy-namespace.yaml",
    "folder": "manifests",
    "manifest_source": "user"
  },
  {
    "file_name": "50-masters-chrony-configuration.yaml",
    "folder": "openshift",
    "manifest_source": "system"
  },
  {
    "file_name": "50-workers-chrony-configuration.yaml",
    "folder": "openshift",
    "manifest_source": "system"
  }
]
{code}

File: internal/versions/common.go [hunk 1]
```diff
@@ -128,19 +129,18 @@ func validateReleaseImageForRHCOS(
 ) error {
 	// Multi is not a valid RHCOS CPU architecture, its sub-architectures are
 	if cpuArchitecture == common.MultiCPUArchitecture {
-		return errors.Errorf("The requested RHCOS version (%s, arch: %s) does not have a matching OpenShift release image", rhcosVersion, cpuArchitecture)
+		return errors.Errorf("The requested RHCOS version (%s, arch: %s) does not have a matching OpenShift release image, there are no valid Openshift versions for architecture %s", rhcosVersion, cpuArchitecture, common.MultiCPUArchitecture)
```
Comment (danmanor) [inline internal/versions/common.go:132]: I would change this message to something like - `RHCOS CPU architecture should be single arch`

File: internal/versions/common.go [hunk 2]
```diff
@@ -160,9 +160,15 @@ func validateReleaseImageForRHCOS(
 				}
 			}
 		}
+		if swag.ContainsStrings(releaseImage.CPUArchitectures, cpuArchitecture) && !swag.ContainsStrings(availableOCPVersions, swag.StringValue(minorVersion)) {
+			availableOCPVersions = append(availableOCPVersions, swag.StringValue(releaseImage.OpenshiftVersion))
+		}
 	}
-
-	return errors.Errorf("The requested RHCOS version (%s, arch: %s) does not have a matching OpenShift release image", *rhcosVersionPtr, cpuArchitecture)
+	ocpVersionAdvice := fmt.Sprintf("There are no OCP versions available in release images for arch %s.", cpuArchitecture)
```
Comment (danmanor) [inline internal/versions/common.go:167]: This function validates that there is a OCP release image that matches the given RHCOS version/arch, I don't think its job is to give the user advice. I would give the advice probably form `RegisterInfraEnvInternal` or `validateArchitectureAndVersion`

File: internal/versions/common.go [hunk 3]
```diff
@@ -160,9 +160,15 @@ func validateReleaseImageForRHCOS(
...
+	if len(availableOCPVersions) > 0 {
+		ocpVersionAdvice = fmt.Sprintf("The following OCP versions are available in release images for arch %s please choose your OCP version to match one of these: %s", cpuArchitecture, strings.Join(availableOCPVersions, ","))
```
Comment (danmanor) [inline internal/versions/common.go:169]: I would change it to something like - `These are the OCP versions which a matching release image has been found for them`

Comment (danmanor) [inline internal/versions/common.go:169]: on premise you can add more release images as well, so I would probably add this option as well

Comment (paul-maidment) [reply internal/versions/common.go:169]: How about `The following OCP versions are available in release images for arch %s please review your release images or choose your OCP version to match one of these` ?

File: internal/versions/common.go (hunk 2 above)
Comment (paul-maidment) [reply internal/versions/common.go:167]: I disagree..I think that the knowledge of whether we have a matching image comes from this area and we only care to advise the user when there is no release image.

Comment (paul-maidment) [reply internal/versions/common.go:167]: I actually disagree.

It's true that `validateInfraEnvCreateParams` makes a call to `validateArchitectureAndVersion`, so you could provide some mechanism through which you provide "advice" when you receive an error from this path. 

It should however be noted that we don't know the real cause of the error as `validateArchitectureAndVersion` can return an error in a couple of scenarios.

To assume that we should provide advice on any error is incorrect.

It is in fact simpler to add the advice here where we have full access to all the facts and we waste no further CPU cycles obtaining it.

Comment (danmanor) [reply internal/versions/common.go:167]: My argument is that this function can potentially be used outside of the infraenv scope, which therefore will be wrong to display such an error. why not add some error matching mechanism one or two levels up the call stack, where the context is clear ?

File: internal/versions/common.go (hunk 3 above)
Comment (danmanor) [reply internal/versions/common.go:169]: Looks good

File: internal/versions/common.go (hunk 2 above)
Comment (paul-maidment) [reply internal/versions/common.go:167]: I guess that at the infraenv scope, we could have an explicit validation to enforce this?

I think we would still have to just create a redundant validation as error matching on the text of the error message? That seems to be a bit of a bad practice surely?

Realistically, would this be called outside of the infraenv scope? Does that happen already?

**Review by danmanor — APPROVED** [review]

**Pull request merged**